tests/unit/steps/test_steps_plan.py

```diff
@@ -638,4 +582,219 @@ def test_verify_data_plane_cli_azs_raise_cannot_find():
     mock_analyze.machines = {"1": mock_machine}
     mock_analyze.data_plane_machines = {"1": mock_machine}
     with pytest.raises(DataPlaneMachineFilterError, match=exp_error_msg):
-        verify_data_plane_cli_azs({"zone-1"}, mock_analyze)
+        cou_plan.verify_data_plane_cli_azs({"zone-1"}, mock_analyze)
+
+
+@pytest.mark.parametrize(
+    "force, novas_machines_input, novas_machines_expected",
+    [
+        (False, [0, 1, 2], [0, 1]),
+        (False, [2], []),
+        (False, [0, 1], [0, 1]),
+        (False, [0], [0]),
+        (True, [0, 1, 2], [0, 1, 2]),
+        (True, [2], [2]),
+        (True, [0], [0]),
+    ],
+)
+@pytest.mark.asyncio
+@patch("cou.steps.plan._get_hypervisors_machines_possible_to_upgrade")
+async def test_filter_hypervisors_machines_by_machine_id(
```

Comment:
This test should test only the `filter_hypervisors_machines` by simulating all type of inputs from user, e.g. `--hosntame bla`, `--machine 1,2` and mock the `_get_hypervisors_machines_possible_to_upgrade` function.

Like that, we will actually test only the functionality of that single function instead of multiple function at once + we should avoid using Nova machines, you should be able to use simple mocks.